### PR TITLE
feat: add CLI config file

### PR DESCRIPTION
**Pull request by jimezesinachi**
# SynthQL pull request template

## Why

Motivation an context for the change.

## What changed

Brief summary of the changes.

## Important

Mention important points that require special attention or that might surprise
reviewers.

### CODE REVIEW

File: packages/cli/src/commands/generate/index.ts [hunk 1]
```diff
@@ -15,14 +17,44 @@ export const generateSchema = async ({
...
+    if (schemaDefOverrides === undefined) {
+        try {
+            return await generate({
+                defaultSchema,
+                connectionString,
+                includeSchemas: schemas,
+                includeTables: tables,
+                outDir: path.resolve(path.join(process.cwd(), out)),
+            });
+        } catch (error) {
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:31]: Don't catch an error just to log it. This isn't really doing anything. Just let the error bubble up.

Comment (fhur) [reply packages/cli/src/commands/generate/index.ts:31]: Same for the other try/catch blocks in this file.

File: packages/cli/src/commands/generate/index.ts [hunk 2]
```diff
@@ -15,14 +17,44 @@ export const generateSchema = async ({
...
+            `${JSON.stringify(schemaDefOverrides, null, 2)}`,
+        ];
 
-    return result;
+        throw new Error(lines.join('\n'));
+    }
 };
+
+function isValidSchemaDefOverrides(value: any): value is SchemaDefOverrides {
+    return value satisfies SchemaDefOverrides;
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:59]: this function is not doing anything at runtime. This function is literally equivalent to 
```ts
function isValidSchemaDefOverrides(x) {return x}
```

File: packages/cli/src/cli.ts [hunk 3]
```diff
@@ -47,22 +47,35 @@ export function cli({
...
                     schemas: argv.schemas,
                     tables: argv.tables,
+                    schemaDefOverrides: argv.schemaDefOverrides,
                 });
-                const tables = Object.keys(schema.properties ?? {});
-                console.warn(`Generated schema with ${tables.length} table(s)`);
-                for (const table of tables) {
-                    console.warn(`- ${table}`);
+
+                if (result) {
```
Comment (fhur) [inline packages/cli/src/cli.ts:69]: This if isn't needed: this function should always return something or throw: see my point further down about not handling the try/catches if you can't actually handle them.

File: packages/cli/src/commands/generate/index.ts [hunk 4]
```diff
@@ -15,14 +17,44 @@ export const generateSchema = async ({
...
+                includeSchemas: schemas,
+                includeTables: tables,
+                outDir: path.resolve(path.join(process.cwd(), out)),
+            });
+        } catch (error) {
+            console.log('Error: ', JSON.stringify(error, null, 2));
+        }
+    } else if (isValidSchemaDefOverrides(schemaDefOverrides)) {
+        try {
+            return await generate({
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:36]: you don't need to calls to `generate`, you should call it just once and always pass the overrides, which can possibly be empty.

File: packages/introspect/src/generate.ts [hunk 5]
```diff
@@ -136,13 +154,25 @@ function createTableJsonSchema(table: TableDetails): JSONSchema {
...
                 ].join('\n'),
             properties: {
-                type: { $ref: `#/$defs/${createTypeDefId(column.type)}` },
-                // a constant value of true
-                selectable: { type: 'boolean', const: true },
-                includable: { type: 'boolean', const: true },
-                whereable: { type: 'boolean', const: true },
-                nullable: { type: 'boolean', const: column.isNullable },
-                isPrimaryKey: { type: 'boolean', const: column.isPrimaryKey },
+                type: columnSchemaOverrides?.type
```
Comment (fhur) [inline packages/introspect/src/generate.ts:157]: Did you try doing this? 

```ts
           {
                type: { $ref: `#/$defs/${createTypeDefId(column.type)}` },
                // a constant value of true
                selectable: { type: 'boolean', const: true },
                includable: { type: 'boolean', const: true },
                whereable: { type: 'boolean', const: true },
                nullable: { type: 'boolean', const: column.isNullable },
                isPrimaryKey: { type: 'boolean', const: column.isPrimaryKey },
                ...overrides
            }

File: packages/introspect/src/generate.ts [hunk 6]
```diff
@@ -244,12 +276,21 @@ function fullTableName(table: TableDetails, defaultSchema: string) {
...
     tables: TableDetails[],
     defaultSchema: string,
+    schemaDefOverrides: SchemaDefOverrides | undefined,
 ): Record<string, JSONSchema> {
     const empty: Record<string, JSONSchema> = {};
 
     return tables.reduce((acc, table) => {
-        acc[createTableDefId(table, defaultSchema)] =
-            createTableJsonSchema(table);
+        const tableSchemaOverrides = schemaDefOverrides
```
Comment (fhur) [inline packages/introspect/src/generate.ts:284]: you don't need a call to `.includes`. This should be a simple call to `schemaDefOverride[table]`

File: packages/introspect/src/generate.ts [hunk 7]
```diff
@@ -186,12 +216,14 @@ function createRootJsonSchema(
     {
         defaultSchema,
         includeTables,
+        schemaDefOverrides,
     }: {
         defaultSchema: string;
         includeTables: string[];
+        schemaDefOverrides: SchemaDefOverrides | undefined;
```
Comment (fhur) [inline packages/introspect/src/generate.ts:223]: instead of passing undefined around all the time, you can just `?? {}` at the top level.

**Review by fhur — CHANGES_REQUESTED** [review]
Good progress jim 👍 

A few things can be simplified but otherwise looks pretty much what I expected.

File: packages/cli/src/cli.ts [hunk 8]
```diff
@@ -47,22 +47,35 @@ export function cli({
                         description:
                             'List of tables to be included in the generation',
                     })
-                    .array('tables');
+                    .array('tables')
+                    .config('configFile')
+                    .parserConfiguration({
```
Comment (fhur) [inline packages/cli/src/cli.ts:52]: what is this for?

File: packages/cli/src/tests/synthql.config.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { CliConfig } from '../types/CliConfig';
+
+export const config: CliConfig = {
+    out: 'src/generated/synthql',
+    defaultSchema: 'public',
+    schemas: ['public'],
+    tables: ['actor', 'customer'],
+    schemaDefOverrides: {
```
Comment (fhur) [inline packages/cli/src/tests/synthql.config.ts:8]: did you check if the overrides will work when you have two tables with the same name defined in different schemas?

Comment (fhur) [reply packages/cli/src/tests/synthql.config.ts:8]: To fix I suggest you require schema.table notation (e.g. public.actor). Its slightly more verbose, but more explicit.

File: packages/introspect/src/generate.ts (hunk 5 above)
Comment (fhur) [reply packages/introspect/src/generate.ts:157]: Much simpler, and scalable implementation.

File: packages/cli/src/cli.ts [hunk 10]
```diff
@@ -47,22 +47,35 @@ export function cli({
...
-                const tables = Object.keys(schema.properties ?? {});
-                console.warn(`Generated schema with ${tables.length} table(s)`);
-                for (const table of tables) {
-                    console.warn(`- ${table}`);
+
+                if (result) {
+                    const tables = Object.keys(result.schema.properties ?? {});
+
+                    console.warn(
+                        `Generated schema with ${tables.length} table(s)`,
```
Comment (fhur) [inline packages/cli/src/cli.ts:73]: You should provide some basic feedback to the user in case the overrides are passed but not actually applied to anything.

File: packages/cli/src/cli.ts (hunk 8 above)
Comment (fhur) [reply packages/cli/src/cli.ts:52]: Ping @jimezesinachi

File: packages/cli/src/cli.ts [hunk 11]
```diff
@@ -47,20 +47,32 @@ export function cli({
...
                     connectionString: argv.connectionString,
                     out: argv.out,
                     defaultSchema: argv.defaultSchema,
                     schemas: argv.schemas,
                     tables: argv.tables,
+                    schemaDefOverrides: argv.schemaDefOverrides,
                 });
-                const tables = Object.keys(schema.properties ?? {});
+
+                // TODO: We want to check if there were any overrides passed
```
Comment (fhur) [inline packages/cli/src/cli.ts:69]: This should be implemented in this MR. It's a very simple check.

Just check if the generated tables include the overrides.

Comment (jimezesinachi) [reply packages/cli/src/cli.ts]: Done now.. I added the comment so that I'll remember to add it this morning

File: packages/cli/src/helpers/getUnappliedSchemaOverridesTableNames.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { isObject } from './isValidSchemaDefOverrides';
+
+export function getUnappliedSchemaOverridesTableNames(
```
Comment (fhur) [inline packages/cli/src/helpers/getUnappliedSchemaOverridesTableNames.ts:3]: remember what I said about co-location: this function is in a "helpers" folder, but it will only ever be used in a single place. So better to co-locate it.

Also in general, try to avoid "helper" or "util" folders. Some things are actually helper/utils but quite often it's just a lazy way to drop unrelated code. So helper folders become a noisy bag of unrelated functionality over time.

File: packages/cli/src/tests/synthql.config.json [hunk 13]
```diff
@@ -16,6 +16,22 @@
             "last_name": {
                 "includable": { "type": "boolean", "const": false }
             }
+        },
+        "public.niggas": {
```
Comment (fhur) [inline packages/cli/src/tests/synthql.config.json]: I don't know jim... I feel like we shouldn't use this word in an open source repo 😅

File: packages/cli/src/helpers/getUnappliedSchemaOverridesTableNames.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    tables: string[],
+    schemaDefOverrides: unknown,
```
Comment (fhur) [inline packages/cli/src/helpers/getUnappliedSchemaOverridesTableNames.ts:5]: You shuold first validate this type, using the JSON schema approach we mentioned earlier and then pass the right type here.

File: packages/cli/src/helpers/isValidSchemaDefOverrides.ts [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+const columnDefPropertyKeys = Object.keys(columnDefProperties);
+
+export function isValidSchemaDefOverrides(
+    value: unknown,
+): value is SchemaDefOverrides | undefined {
+    if (isObject(value)) {
+        for (const [tableDefName, tableDefOverride] of Object.entries(value)) {
+            if (containsSingleDot(tableDefName)) {
```
Comment (fhur) [inline packages/cli/src/helpers/isValidSchemaDefOverrides.ts:19]: Jim why didn't you go with the JSON schema approach?

https://discord.com/channels/1222945509266817055/1222945509740642317/1263412303445889044

Comment (fhur) [reply packages/cli/src/helpers/isValidSchemaDefOverrides.ts:19]: Also I feel like you haven't really "grokked" the idea of returning early

Comment (fhur) [reply packages/cli/src/helpers/isValidSchemaDefOverrides.ts:19]: Here is a sample schema btw:

```json
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "properties": {
        "tables": {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        "schemas": {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        "defaultSchema": {
            "type": "string",
        },
        "schemaDefOverrides": {....}
    },
    "required": ["tables", "schemaDefOverrides"]
}
```

File: packages/cli/src/tests/synthql.config.json (hunk 13 above)
Comment (jimezesinachi) [reply packages/cli/src/tests/synthql.config.json]: I agree.. removed already

File: packages/cli/src/commands/generate/index.ts [hunk 16]
```diff
@@ -1,12 +1,15 @@
-import { generate } from '@synthql/introspect';
 import path from 'path';
+import { generate } from '@synthql/introspect';
+import { isValidSchemaDefOverrides } from '../../validators/isValidSchemaDefOverrides';
+import { SchemaDefOverrides } from '@synthql/queries';
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:4]: SchemaDefOverrides is a CLI-only construct. It should therefore be defined in the CLI package.

File: packages/cli/src/commands/generate/index.ts [hunk 17]
```diff
@@ -15,14 +18,14 @@ export const generateSchema = async ({
...
 }: GenerateSchemaOptions) => {
-    const result = await generate({
+    return await generate({
         defaultSchema,
         connectionString,
         includeSchemas: schemas,
         includeTables: tables,
-        outDir: out,
+        schemaDefOverrides: schemaDefOverrides,
+        outDir: path.resolve(path.join(process.cwd(), out)),
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:29]: jim you're not paying attention: this resolve/join/cwd pattern is buggy, it doesn't respect absolute paths.

Consider what happens when you use the absolute path "/tmp/foo.ts", it will be converted into ./tmp/foo.ts, which is incorrect.

I've mentioned this error in the past, please don't introduce this bug again.

File: packages/cli/src/commands/generate/index.ts (hunk 16 above)
Comment (jimezesinachi) [reply packages/cli/src/commands/generate/index.ts:4]: That is not correct 😅... It is also used in the introspect package

File: packages/cli/src/validators/schemas.ts [hunk 18]
```diff
@@ -0,0 +1,341 @@
+import { JSONSchemaType } from 'ajv';
+import { CliConfig } from '../types/CliConfig';
+
+export const configFileSchema: JSONSchemaType<Partial<CliConfig>> = {
```
Comment (fhur) [inline packages/cli/src/validators/schemas.ts:4]: there is a LOT of duplication in this file.

I would suggest splitting out a `tableSchema` and adding a function to make some fields nullable, similar to typescripts `Nullable<T>` type.

File: packages/cli/src/validators/schemas.ts [hunk 19]
```diff
@@ -0,0 +1,341 @@
...
+            nullable: true,
+            required: [],
+            additionalProperties: false,
+        },
+    },
+    required: [],
+    additionalProperties: false,
+};
+
+export const cliArgvSchema: JSONSchemaType<CliConfig> = {
```
Comment (fhur) [inline packages/cli/src/validators/schemas.ts:175]: I'm not sure why this is needed 🤔 doesn't `yargs` provide type-safety already?

File: packages/queries/src/types/Schema.ts [hunk 20]
```diff
@@ -47,38 +47,46 @@ export interface ColumnDef {
...
-        };
-        nullable: {
-            type: string;
-            const: boolean;
-        };
-        isPrimaryKey: {
-            type: string;
-            const: boolean;
-        };
+        type: ColumnDefTypeDef;
```
Comment (fhur) [inline packages/queries/src/types/Schema.ts:50]: I'm not sure this is an improvement in readability over the previous version.

`{type:string, const:boolean}` is much easier to read than `ColumnDefAttributeDef`. Seems like a layer of indirection to me.

File: packages/introspect/src/generate.ts [hunk 21]
```diff
@@ -116,10 +148,25 @@ function createTableDefId(type: TableDetails, defaultSchema: string) {
...
 }
 
-function createTableJsonSchema(table: TableDetails): JSONSchema {
+function createTableJsonSchema(
+    table: TableDetails,
+    tableDefOverrides: TableDefOverrides | undefined,
+): JSONSchema {
     const empty: Record<string, any> = {};
 
+    // Here, we want to check if there was overriding table schema def
```
Comment (fhur) [inline packages/introspect/src/generate.ts:157]: I think you're missing an `applyOverrides` function that encapsulates this logic.

File: packages/cli/src/cli.ts [hunk 22]
```diff
@@ -47,22 +50,76 @@ export function cli({
...
-                });
-                const tables = Object.keys(schema.properties ?? {});
-                console.warn(`Generated schema with ${tables.length} table(s)`);
-                for (const table of tables) {
-                    console.warn(`- ${table}`);
+                // We do this so that we lose the types from yargs but we gain
+                // complete type-safety after we pass the data through the validator
+                const data = { ...argv, connectionString: undefined } as any;
+
+                if (validateCliArgv(data)) {
```
Comment (fhur) [inline packages/cli/src/cli.ts:81]: Please use the return early pattern. This is much harder to read this way. It breaks being able to linearly scan the text.

File: packages/cli/src/cli.ts [hunk 23]
```diff
@@ -47,22 +50,76 @@ export function cli({
...
-                    schemas: argv.schemas,
-                    tables: argv.tables,
-                });
-                const tables = Object.keys(schema.properties ?? {});
-                console.warn(`Generated schema with ${tables.length} table(s)`);
-                for (const table of tables) {
-                    console.warn(`- ${table}`);
+                // We do this so that we lose the types from yargs but we gain
+                // complete type-safety after we pass the data through the validator
+                const data = { ...argv, connectionString: undefined } as any;
```
Comment (fhur) [inline packages/cli/src/cli.ts:79]: Don't use `as any`. For cases like these use `as unknown`. 

`as any` is dangerous, `as unknown` is safe.
`as any` allows any compiler errors because it allows any type.
`as unknown` will force you to check for the type, so it's totally safe to use it.

File: packages/cli/src/commands/generate/index.ts (hunk 16 above)
Comment (fhur) [reply packages/cli/src/commands/generate/index.ts:3]: Yeah about that: we also need to refactor that. We shouldn't couple the  /introspect package to the CLI. I'm about to post a comment around that.

File: packages/cli/src/commands/generate/index.ts [hunk 24]
```diff
@@ -15,14 +17,14 @@ export const generateSchema = async ({
...
+    schemaDefOverrides,
 }: GenerateSchemaOptions) => {
-    const result = await generate({
+    return await generate({
         defaultSchema,
         connectionString,
         includeSchemas: schemas,
         includeTables: tables,
-        outDir: out,
+        schemaDefOverrides: schemaDefOverrides,
```
Comment (fhur) [inline packages/cli/src/commands/generate/index.ts:27]: SchemaDefOverrides is a `CLI`-only construct. It shouldn't "leak" into the /introspect package.

We currently only allow synthql.cofig.json, but in the near future we will also allow synthql.config.(js|ts), where we will allow users to maybe functions to override the schemas.

The important bit to take from this is that schemaDefOverrides as a concept is very much dependent on the shape of the synthql.config.json and we shouldn't leak that implementation detail into the /introspect pacakge.

As you also noticed, this also forces us to share a type between these two packages through the /queries package (which has nothing to do with this at all).

So what's the proper fix? Quite simple, actually. See the following pseudocode:

1. Add a property on the generate function that allows for passing an array of table overriders.
```ts
Array<{
    test: (table) => boolean,
    applyOverride: (oldTable) => NewTable
}>
```
These overriders can specify a `test` to determine on which tables they should apply to, and when the test is true, a function that applies the override.

2. Add a function that takes `synthql.config.json` format and creates an array of overriders 
```ts
const overrides = createTableOverrides(jsonConfig)
```

File: packages/cli/src/validators/schemas.ts (hunk 18 above)
Comment (jimezesinachi) [reply packages/cli/src/validators/schemas.ts:4]: This is what I originally tried to do, but for type breaks whenever you assign a passed variable to the JSONSchemaType

File: packages/cli/src/validators/schemas.ts (hunk 19 above)
Comment (jimezesinachi) [reply packages/cli/src/validators/schemas.ts:175]: It does, but because schemaDefOverrides is read from the config file, it's type is unknown to yargs, so we use that to give the values their correct type

File: packages/cli/src/commands/generate/index.ts (hunk 16 above)
Comment (fhur) [reply packages/cli/src/commands/generate/index.ts:3]: See https://github.com/synthql/SynthQL/pull/29#discussion_r1686162734

File: packages/cli/src/validators/schemas.ts (hunk 19 above)
Comment (fhur) [reply packages/cli/src/validators/schemas.ts:175]: Ok fair, but that means we just need a schema for the schemaDefOverrides, not for all the other args, right?

Comment (jimezesinachi) [reply packages/cli/src/validators/schemas.ts:175]: That is correct... I will adjust it now 👍🏾

File: packages/queries/src/types/Schema.ts (hunk 20 above)
Comment (jimezesinachi) [reply packages/queries/src/types/Schema.ts:50]: Wasn't much for readablity as it was for reusability (i.e. deduplication)... but if it's severely affecting readability, as can switch it back

Comment (fhur) [reply packages/queries/src/types/Schema.ts:50]: De-duplication between the fields?

Comment (jimezesinachi) [reply packages/queries/src/types/Schema.ts:50]: Yes.. the selectable, whereable, includeable, nullable and isPrimaryKey fields are all assigned to an identical object: `{ type:string, const:boolean }`, which is what `ColumnDefAttributeDef` is equal to

File: packages/cli/src/e2e.test.ts [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+
+            // Verify that the passed attribute overrides were applied
+            const schemaFilePath = path.join(outDir, `schema.ts`);
+
+            const file = await import(schemaFilePath);
+            const schema = file.schema.properties;
+
+            const schemaDefOverrides = output.schemaDefOverrides as any;
+
+            for (const qualifiedTableName in schemaDefOverrides) {
```
Comment (fhur) [inline packages/cli/src/e2e.test.ts:107]: wow what's happening here? what are you trying to assert?

File: packages/cli/src/e2e.test.ts [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+            // Verify that the option data is correct
+            expect(output.configFile).toEqual(configFile);
+
+            // Verify that the passed attribute overrides were applied
+            const schemaFilePath = path.join(outDir, `schema.ts`);
+
+            const file = await import(schemaFilePath);
+            const schema = file.schema.properties;
+
+            const schemaDefOverrides = output.schemaDefOverrides as any;
```
Comment (fhur) [inline packages/cli/src/e2e.test.ts:105]: Please don't use `any`. Is there a reason why you need it here?

File: packages/cli/src/e2e.test.ts (hunk 25 above)
Comment (fhur) [reply packages/cli/src/e2e.test.ts:107]: Usually it's not a good idea to have too much logic inside your tests, as bugs can creep into the logic over time.

Can we simplify this logic? Why can't we just do a single big `expect`?

File: packages/cli/src/tests/synthql.dev.config.json [hunk 27]
```diff
@@ -0,0 +1,179 @@
+{
```
Comment (fhur) [inline packages/cli/src/tests/synthql.dev.config.json:1]: what is the purpose of this file?

File: packages/cli/src/types/CliConfig.ts [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+     * The table definitions for any tables you want
+     * to be overriden during the schema generation
+     */
+    schemaDefOverrides?: SchemaDefOverrides;
+}
+export interface SchemaDefOverrides {
+    /**
+     * The column definitions for each table to be overriden
+     * The key for each table should be the qualified name for the table,
+     * in the form `${schemaName.tableName}` e.g `'public.users'`
```
Comment (fhur) [inline packages/cli/src/types/CliConfig.ts:36]:
```suggestion
     * in the form `${schemaName}.${tableName}` e.g `'public.users'`
```

File: packages/cli/src/types/CliConfig.ts [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    [x: string]: TableDefOverrides | undefined;
```
Comment (fhur) [inline packages/cli/src/types/CliConfig.ts:38]: I would suggest we don't introduce `TableDefOverrides`. Also that we give `x` a better name.

```suggestion
    [table: string]: { [column: string]: Partial<ColumnDefProperties> }
```

File: packages/cli/src/types/CliConfig.ts [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+}
+export interface TableDefOverrides {
```
Comment (fhur) [inline packages/cli/src/types/CliConfig.ts:40]: See my suggestion above to remove this. As a general rule of thumb, try to keep the number of names you introduce into the codebase as small as possible.

Just because something has a name doesn't mean that the name is more descriptive than the thing being named.

For example `T | null` is much more readable and descriptive than `Option<T>`. There's two reasons for this:
1. `T | null` will be familiar to every single TS/JS dev in the world, unlike `Option<T>` which could be implemented as `T | null` or `T | undefined` or `T | null | undefined`
2. A name's usefulness increases in proportion to how often the name is used in the codebase. Assuming that Option<T> was implemented as `T | null | undefined` and this pattern was used everywhere in the codebase, then that probably OK. Now consider the `TableDedOverrides` which is used only once (maybe twice?) in the entire codebase. People that look at that name will always have to look at the implementation.

File: packages/cli/src/validators/createTableDefOverriderList.ts [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+import { SchemaDefOverrides } from '../types/CliConfig';
+
+interface TableDefOverrider {
+    isValidQualifiedTableName: (qualifiedTableName: string) => boolean;
+    applyColumnDefOverride: (
+        columnName: string,
+    ) => Partial<ColumnDefProperties> | undefined;
+}
+
+type TableDefOverriderList = Array<TableDefOverrider>;
```
Comment (fhur) [inline packages/cli/src/validators/createTableDefOverriderList.ts:11]: Another example of bad naming. The name is worse than the thing being named. This is just a layer of indirection.

File: packages/cli/src/validators/createTableDefOverriderList.ts [hunk 32]
```diff
@@ -0,0 +1,36 @@
+import { ColumnDefProperties } from '@synthql/queries';
+import { SchemaDefOverrides } from '../types/CliConfig';
+
+interface TableDefOverrider {
+    isValidQualifiedTableName: (qualifiedTableName: string) => boolean;
+    applyColumnDefOverride: (
+        columnName: string,
+    ) => Partial<ColumnDefProperties> | undefined;
```
Comment (fhur) [inline packages/cli/src/validators/createTableDefOverriderList.ts:8]: This shouldn't ever return `undefined`.

File: packages/cli/src/validators/createTableDefOverriderList.ts [hunk 33]
```diff
@@ -0,0 +1,36 @@
+import { ColumnDefProperties } from '@synthql/queries';
+import { SchemaDefOverrides } from '../types/CliConfig';
+
+interface TableDefOverrider {
+    isValidQualifiedTableName: (qualifiedTableName: string) => boolean;
```
Comment (fhur) [inline packages/cli/src/validators/createTableDefOverriderList.ts:5]: "valid" is the wrong adjective here. I would call this `test`, `predicate`, `isApplicable`.

File: packages/cli/src/validators/createTableDefOverriderList.ts [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+    applyColumnDefOverride: (
```
Comment (fhur) [inline packages/cli/src/validators/createTableDefOverriderList.ts:6]: This signature doesn't make sense. This should take an existing `ColumnDefProperties` as input and just transform it.

File: packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts [hunk 35]
```diff
@@ -0,0 +1,33 @@
+import { SchemaDefOverrides } from '../types/CliConfig';
+
+export function getUnappliedSchemaOverridesTableNames(
+    tables: string[],
+    defaultSchema: string,
+    schemaDefOverrides?: SchemaDefOverrides,
+) {
+    const unappliedSchemaOverrideTableKeys: string[] = [];
+
+    if (isObject(schemaDefOverrides)) {
```
Comment (fhur) [inline packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts:10]: you don't need isObject. Simply do 

```
if (!schemaDefOverrides) {
    return []
}
```

File: packages/cli/src/tests/synthql.dev.config.json (hunk 27 above)
Comment (fhur) [reply packages/cli/src/tests/synthql.dev.config.json:1]: nvm, for the tests

File: packages/cli/src/e2e.test.ts [hunk 36]
```diff
@@ -0,0 +1,179 @@
...
+} from 'vitest';
+import * as prettier from 'prettier';
+import fs from 'fs';
+import path from 'path';
+import { configFileSchema } from './validators/schemas';
+import { cli } from './cli';
+import { randomUUID } from 'crypto';
+
+describe('e2e', () => {
+    const out = 'generated/synthql/e2e/tests';
```
Comment (fhur) [inline packages/cli/src/e2e.test.ts:18]: if these are temporary files it would be better to write them to /tmp

File: packages/cli/src/e2e.test.ts [hunk 37]
```diff
@@ -0,0 +1,179 @@
...
+    vitest,
+} from 'vitest';
+import * as prettier from 'prettier';
+import fs from 'fs';
+import path from 'path';
+import { configFileSchema } from './validators/schemas';
+import { cli } from './cli';
+import { randomUUID } from 'crypto';
+
+describe('e2e', () => {
```
Comment (fhur) [inline packages/cli/src/e2e.test.ts:17]: Conceptually, what's the difference between `src/e2e.test.ts` and `src/cli.test.ts`?

File: packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts (hunk 35 above)
Comment (fhur) [reply packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts:10]: In fact, here's a simplified version:

```
import { SchemaDefOverrides } from '../types/CliConfig';

export function findUnusedOverrides(
    existingTables: string[],
    defaultSchema: string,
    schemaOverrides?: SchemaDefOverrides,
): string[] {
    if (!Boolean(schemaOverrides)) {
        return [];
    }

    return Object.keys(schemaOverrides).flatMap((fullTableName) => {
        const [schemaName, tableName] = fullTableName.split('.');
        const isDefaultSchema = schemaName === defaultSchema;
        const tableToCheck = isDefaultSchema ? tableName : fullTableName;

        return !existingTables.includes(tableToCheck) ? [`- ${fullTableName}`] : [];
    });
}
```

File: packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts [hunk 38]
```diff
@@ -0,0 +1,33 @@
...
+        const tablesToOverride = Object.keys(schemaDefOverrides);
+
+        for (const tableKey of tablesToOverride) {
+            const [schemaName, tableName] = tableKey.split('.');
+
+            if (schemaName === defaultSchema) {
+                if (!tables.includes(tableName)) {
+                    unappliedSchemaOverrideTableKeys.push(`- ${tableKey}`);
```
Comment (fhur) [inline packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts:18]: It's really weird that this function returns an array of `- table`. It completely breaks the potential composability of this function.

I would suggest you remove the `-` part.

It's like mixing two concerns into one: the rendering (`- `) and the logic (finding unused overrides).

File: packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts [hunk 39]
```diff
@@ -0,0 +1,33 @@
...
+                }
+            }
+        }
+    }
+
+    return unappliedSchemaOverrideTableKeys;
+}
+
+export function isObject(value: unknown): value is object {
```
Comment (fhur) [inline packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts:31]: Are we using this anywhere else? if not, we should remove it as it's not really needed in this function.

File: packages/cli/src/e2e.test.ts (hunk 37 above)
Comment (jimezesinachi) [reply packages/cli/src/e2e.test.ts:17]: `e2e.test.ts` is just where I'm adding any tests where we're testing the results (and not just the `CLI` function, like the tests in `cli.test.ts`) and/or are storing the side-effects, using it the same way you used it in the introspect package

File: packages/cli/src/tests/synthql.dev.config.json (hunk 27 above)
Comment (jimezesinachi) [reply packages/cli/src/tests/synthql.dev.config.json:1]: It's a copy of the config schema I'm using to validate the version I'm using for the tests

File: packages/cli/src/e2e.test.ts (hunk 26 above)
Comment (jimezesinachi) [reply packages/cli/src/e2e.test.ts:105]: Indeed. Same reason why we needed the` ajv `typing. `schemaDefOverrides` is always `unknown`. I thought of using the validator function (`validateSchemaDefOverrides`)to validate the value, but that would be adding a dependency to the tests. The idea of using any here is that we want this test to fail with a reference error, the moment the data shape is not what it expects, without having to use a function from a dependency

File: packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts (hunk 38 above)
Comment (fhur) [reply packages/cli/src/validators/getUnappliedSchemaOverridesTableNames.ts:18]: Yeah, for sure. there is essentially _zero_ cost to iterating what will in the worst case be a few 1000 items.

File: packages/cli/src/tests/synthql.dev.config.json (hunk 27 above)
Comment (fhur) [reply packages/cli/src/tests/synthql.dev.config.json:1]: Can you add a comment somewhere (i.e. where you generate it) explaining what this is for

File: packages/cli/src/e2e.test.ts (hunk 25 above)
Comment (jimezesinachi) [reply packages/cli/src/e2e.test.ts:107]: I will look into simplifying it, but a big expect might not be possible because the objects don't have the same shape

File: packages/cli/src/e2e.test.ts (hunk 26 above)
Comment (fhur) [reply packages/cli/src/e2e.test.ts:105]: I think using the validator makes a lot of sense, specially given that this is an e2e test (we're testing the whole flow, including validation). Why would adding a dependency on it be a bad idea?

Comment (jimezesinachi) [reply packages/cli/src/e2e.test.ts:105]: The way I was thinking about it, since we were using the validator to parse the data inside the CLI function, it seemed like testing the output of the validator process using the same validator... but if we're okay with that, then it makes sense to use it

File: packages/cli/src/types/CliConfig.ts [hunk 40]
```diff
@@ -0,0 +1,53 @@
+import { ColumnDefProperties } from '@synthql/queries';
+export interface CliConfig {
+    /**
+     * Internal. DO NOT USE/SET this option.
+     * Specifies which draft of the JSON Schema
+     * standard the schema adheres to.
+     */
+    $schema?: string;
```
Comment (fhur) [inline packages/cli/src/types/CliConfig.ts:8]: instead of string we should probably just spe

**Review by fhur — DISMISSED** [review]

**Review by fhur — APPROVED** [review]
